src/codeu/chat/util/PasswordStorage.java

```diff
@@ -0,0 +1,241 @@
+        } catch (NumberFormatException ex) {
+            throw new InvalidHashException(
+                "Could not parse the hash size as an integer.",
+                ex
+            );
+        }
+
+        if (storedHashSize != hash.length) {
+            throw new InvalidHashException(
+                "Hash length doesn't match stored hash length."
+            );
+        }
+
+        // Compute the hash of the provided password, using the same salt, 
+        // iteration count, and hash length
+        byte[] testHash = pbkdf2(password, salt, iterations, hash.length);
+        // Compare the hashes in constant time. The password is correct if
+        // both hashes match.
+        return slowEquals(hash, testHash);
+    }
+
+    private static boolean slowEquals(byte[] a, byte[] b)
+    {
+        int diff = a.length ^ b.length;
+        for(int i = 0; i < a.length && i < b.length; i++)
+            diff |= a[i] ^ b[i];
+        return diff == 0;
+    }
+
+    private static byte[] pbkdf2(char[] password, byte[] salt, int iterations, int bytes)
+        throws CannotPerformOperationException
+    {
+        try {
+            PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, bytes * 8);
+            SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+            return skf.generateSecret(spec).getEncoded();
+        } catch (NoSuchAlgorithmException ex) {
+            throw new CannotPerformOperationException(
+                "Hash algorithm not supported.", ex);
+        } catch (InvalidKeySpecException ex) {
+            throw new CannotPerformOperationException(
+                "Invalid key spec.", ex);
+        }
+    }
+
+    private static byte[] fromBase64(String hex)
+        throws IllegalArgumentException
+    {
+        return DatatypeConverter.parseBase64Binary(hex);
+    }
+
+    private static String toBase64(byte[] array)
+    {
+        return DatatypeConverter.printBase64Binary(array);
+    }
+
+    /* My added methods */
+
+    public static String setPassword() {
+        Console console = System.console();
+        char password[] = console.readPassword("Enter a new password: ");
+        char passwordConfirm[] = console.readPassword("Retype your password: ");
+        System.out.println("Verifying...");
+        while (!Arrays.equals(password, passwordConfirm)) {
+            System.out.println("Passwords didn't match. Please try again.");
+            password = console.readPassword("Enter a new password: ");
+            passwordConfirm = console.readPassword("Retype your password: ");
+            System.out.println("Verifying...");
+        }
+        try {
+            return PasswordStorage.createHash(password);
+        } catch (PasswordStorage.CannotPerformOperationException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    public static void authenticate(String password) {
+        Console console = System.console();
```

Comment:
Bad-Design-Alert! This is a util function where you are asking the user to input some value. Anything that requires user input should strictly be part of the client. Same old logic: isolate different parts of the application to as few classes as possible.
Better solution: 
1. Read the password in Chat.java before you call this function. 
2. Pass enteredPassword and userId to the Context, which calls Client controller, which makes an authentication request. (Note: Let's discuss the role of Context in our next hangout)
3. Handle the authentication request in the server Controller; retrieve the actual password for the userId supplied, then call PasswordUtils.doPasswordsMatch(enteredPassword, actualPassword). Return a boolean back to the client indicating whether passwords matched.